### PR TITLE
GitHub-Actions: package creation & image building refactors

### PR DESCRIPTION
**Pull request by alltilla**
This PR:
  - refactors the package creation and upload to a reusable workflow
  - fixes an ordering issue in draft-release
  - removes a redundant nightly build for "Binary packages"
  - reorders the dbld image build and the nightly package creation schedules
  - does some renaming in some workflows
  - fixes some typos

---

**Example runs**

Temporary changes:
  * Removed the `push` trigger, so we can see a normal PR run (no duplicated jobs for `pull_request` and `push`).
  * Added `alltilla-test` to Azure upload path for testing.
  * Minimized the macOS workflow, as we do not need that for testing this refactor.

Positive runs:
- [x] PR: https://github.com/alltilla/syslog-ng/pull/97
- [x] Nightly packages: https://github.com/alltilla/syslog-ng/actions/runs/2113818151
- [x] Version bump: https://github.com/alltilla/syslog-ng/pull/98
- [x] Draft release: https://github.com/alltilla/syslog-ng/actions/runs/2113971077

Negative runs:
- [x] Nightly packages: tarball creation fail: https://github.com/alltilla/syslog-ng/actions/runs/2114067344
    * create-packages and upload-packages were skipped
- [x] Nightly packages: create-packages fail: https://github.com/alltilla/syslog-ng/actions/runs/2114189918
    * upload-packages was skipped
- [x] Nightly packages: upload-packages fail: https://github.com/alltilla/syslog-ng/actions/runs/2114530194
- [x] Draft release: job cancel: https://github.com/alltilla/syslog-ng/actions/runs/2114822767
    * cancelled at tarball creation
    * create-packages, upload-packages and create-draft-release were skipped
    * failure was commented on the version bump PR: https://github.com/alltilla/syslog-ng/pull/103
- [x] Draft release: tarball creation fail: https://github.com/alltilla/syslog-ng/actions/runs/2114158609
    * create-packages, upload-packages and create-draft-release were skipped
    * failure was commented on the version bump PR: https://github.com/alltilla/syslog-ng/pull/99
- [x] Draft release: create-packages fail: https://github.com/alltilla/syslog-ng/actions/runs/2114298334
    * upload-packages and create-draft-release were skipped
    * failure was commented on the version bump PR: https://github.com/alltilla/syslog-ng/pull/100
- [x] Draft release: upload-packages fail: https://github.com/alltilla/syslog-ng/actions/runs/2114402843
    * create-draft-release was skipped
    * failure was commented on the version bump PR: https://github.com/alltilla/syslog-ng/pull/101
- [x] Draft release: create-draft-release fail: https://github.com/alltilla/syslog-ng/actions/runs/2114690280
    * failure was commented on the version bump PR: https://github.com/alltilla/syslog-ng/pull/102

Signed-off-by: Attila Szakacs <attila.szakacs@oneidentity.com>

### CODE REVIEW

**Comment by alltilla** [comment]
Question: Should we limit the dbld image creation workflow run to some specific paths in case of `pull_request` and `push`?

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by MrAnno** [comment]
> Question: Should we limit the dbld image creation workflow run to some specific paths in case of pull_request and push?

https://github.com/syslog-ng/syslog-ng/pull/3932#issuecomment-1057867489

**Comment by gaborznagy** [comment]
> > Question: Should we limit the dbld image creation workflow run to some specific paths in case of pull_request and push?
> 
> [#3932 (comment)](https://github.com/syslog-ng/syslog-ng/pull/3932#issuecomment-1057867489)

Is it okay if I add it to #3932 ? We'll have a minimal conflict then.

**Comment by alltilla** [comment]
Rebased to master.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by alltilla** [comment]
Updated:
  * Fixed unnecessary draft release job comment on every PR.
  * Fixed Azure upload path.
  * Fixed secret handling in `upload-packages`.
  * Added `name:` to `create-packages` and `upload-packages`, to remove job name duplication.

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: .github/workflows/draft-release.yml [hunk 1]
```diff
@@ -61,85 +72,61 @@ jobs:
...
             --file /tmp/message \
-            --attach ${TARBALL_PATH} \
-            ${RELEASE_TAG}
+            --attach ${{ needs.create-release-tarball.outputs.TARBALL_NAME }} \
+            ${{ needs.create-release-tarball.outputs.RELEASE_TAG }}
 
+  comment-workflow-result:
+    runs-on: ubuntu-latest
+    needs: create-draft-release
+    if: always() && github.event_name == 'pull_request' && needs.create-draft-release.result != 'skipped'
```
Comment (alltilla) [inline .github/workflows/draft-release.yml:119]: Note to self:

Maybe instead of `needs.create-draft-release.result != 'skipped'`, `needs.create-release-tarball != 'skipped'` would be better. I think the `create-draft-release` job can be skipped, if a previous job failed, and in this case we want to comment the failure.

**Review by gaborznagy — COMMENTED** [review]
I really like this refactor!

Especially that instead of conditionals to upload a package, you use inputs!

File: .github/workflows/draft-release.yml [hunk 2]
```diff
@@ -61,45 +72,61 @@ jobs:
...
+    needs: create-packages
+    uses: ./.github/workflows/upload-packages.yml
+    with:
+      pkg-type: stable
+    secrets:
+      azure-sas-token: ${{ secrets.AZURE_SAS_TOKEN }}
+
+  create-draft-release:
+    runs-on: ubuntu-latest
+    needs: [create-release-tarball, upload-packages]
```
[the remark this answers is not in the document]
Comment (alltilla) [reply .github/workflows/draft-release.yml:96]: Yes, I noticed the hover thing, too.

`create-release-tarball` is needed, because we need `outputs` from there, `upload-packages` is needed, so `create-draft-release` does not start until it is finished.

I can add every other step, too, so the hovering is better, but I would prefer the simplicity of the workflow file over the mouse hovering "fix".

File: .github/workflows/create-packages.yml [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+      - name: Create package
+        working-directory: syslog-ng
+        run: |
+          ./dbld/rules package-${{ matrix.distro }}
+
+      - name: Prepare package for artifact
+        run: |
+          mkdir package
+          cp -r syslog-ng/dbld/build/${{ matrix.distro }} package/
```
[the remark this answers is not in the document]
Comment (alltilla) [reply .github/workflows/create-packages.yml:75]: I think it is necessary, but maybe I am wrong.

`upload-artifact` "flattens" the path, so `path: /home/alltilla/whatever` would upload the contents of `whatever`, but not the directory itself.

I wanted to keep the `whatever` directory in the artifact, but `/home/alltilla` has other directories as well, so I could not write `/home/alltilla/*`, so I created a new, empty directory (e.g. `/home/alltilla/foo`), and copied `whatever` there, so `path: /home/alltilla/foo/*` will upload the `whatever` directory with its contents.

I would prefer if it was not needed, maybe others can find a simpler solution. I will check it again, though!

File: .github/workflows/create-packages.yml [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+      - name: Store package as artifact
+        uses: actions/upload-artifact@v3
+        with:
+          name: package-${{ matrix.distro }}
```
[the remark this answers is not in the document]
Comment (alltilla) [reply .github/workflows/create-packages.yml:80]: No we don't, it was just a preference from me. The artifacts on the GUI look nice this way, but I am open to change it.

File: .github/workflows/draft-release.yml [hunk 5]
```diff
@@ -44,15 +52,18 @@ jobs:
...
           . .github/workflows/gh-tools.sh
 
           VERSION=`cat VERSION`
           RELEASE_TAG=syslog-ng-$VERSION
           RELEASE_NAME=${RELEASE_TAG}
-          TARBALL_PATH=dbld/release/${VERSION}/${RELEASE_NAME}.tar.gz
+          TARBALL_NAME=${RELEASE_NAME}.tar.gz
+          TARBALL_PATH=dbld/release/${VERSION}/${TARBALL_NAME}
 
           gh_export VERSION RELEASE_TAG RELEASE_NAME TARBALL_PATH
```
[the remark this answers is not in the document]
Comment (alltilla) [reply .github/workflows/draft-release.yml:65]: Sure, I will check it!

File: .github/workflows/draft-release.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (gaborznagy) [reply .github/workflows/draft-release.yml:96]: Maybe I was unclear, sorry. Let me try again.
(of course this is minor thing, but I think it worth it as I got confused about the graph of the jobs)

Jobs in a workflow run parallel by default, you can make them run in a sequence with the `needs` keyword.
Draft-release is now made out of multiple jobs, and they are ordered in a strict sequence.

What I meant earlier is that it should be enough that 1 job only depends on the preceding job.
It's not needed to mention earlier jobs to get their artifacts, AFAIK artifacts are shared in a workflow run (and this is different to GitLab CI I think).
It's not "just a UI thing", but it can be confusing to understand the order in which the jobs run: `needs` show what follows what.
Seeing multiple jobs in a job's `needs` keyword caught my eye as I suddenly got confused about what's running concurrently.

[the remark this answers is not in the document]
Comment (alltilla) [reply .github/workflows/draft-release.yml:96]: `create-release-tarball` is not needed because of its artifact, but because of its `outputs`, see `create-draft-release` -> `Create draft release`, e.g.: `${{ needs.create-release-tarball.outputs.RELEASE_TAG }}`.

I would also be happy, if I could only reference the previous job, but unfortunately it is not possible in this case.

File: .github/workflows/create-packages.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (gaborznagy) [reply .github/workflows/create-packages.yml:75]: Thanks for the explanation!
I've checked our current package uploads (e.g. [here](https://github.com/syslog-ng/syslog-ng/actions/runs/2000844525) is a run), and this one, and yes, the directory is not kept.

I don't think it's needed to create unique directories for artifacts, that's solved by Github Actions if you give a **unique name** for artifact upload.
Maybe I don't understand the motivation behind keeping the directory. You can avoid overwriting previously downloaded artifacts with a unique `path` when downloading.

[the remark this answers is not in the document]
Comment (gaborznagy) [reply .github/workflows/create-packages.yml:75]: Okay, I found where is the directory name of the packages (e.g. "centos-7) needed (at leas now):
In `upload-packages.yml`, the job first downloads the given package by the name `package-centos-7` and then when it's about to upload the packages to Azure, it **knows** that there is a `centos-7` folder (by using `${{ matrix.distro }}` context variable.
It's confusing that the distro variable is used for the source parameter: it is not related to the list in the `distros` input variable, it is a knowledge from the `create-packages` job that a certain directory will be created.
I wouldn't know it internal knowledge though, as you can expect the content of an artifact, but if we would change something in `create-packages` you could break `upload-packages` if it's not tested.
In my opinion it would be easier if we add a path to the `download-artifact` step and then use that directory for uploading to azure.

File: .github/workflows/create-packages.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (gaborznagy) [reply .github/workflows/create-packages.yml:80]: It's okay this way.

File: .github/workflows/draft-release.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (gaborznagy) [reply .github/workflows/draft-release.yml:96]: Ohhh, my bad. You even emphasized the reason is `outputs`.... Sorry I just didn't parse your whole comment at first.
Thanks for clarifying.

File: .github/workflows/draft-release.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (alltilla) [reply .github/workflows/draft-release.yml]: Thanks, fixed!

File: .github/workflows/create-packages.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (alltilla) [reply .github/workflows/create-packages.yml:75]: Good idea, thanks. I am trying to make the paths more simpler with your idea.

File: .github/workflows/draft-release.yml (hunk 1 above)
Comment (alltilla) [reply .github/workflows/draft-release.yml]: Good idea, thanks.

I used `${{ success() }}` and `${{ cancelled() }}` in the commenting instead.

I prefer the `always()` + `pull_request` contidions, because `always()` runs in case of `cancelled()`, too. I also added a comment for the `cancelled()` result.

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: .github/workflows/draft-release.yml (hunk 1 above)
Comment (alltilla) [reply .github/workflows/draft-release.yml]: `${{ success() }}` and `${{ cancelled() }}` does not work in the script :( I falled back to `result`...

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: .github/workflows/create-packages.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (alltilla) [reply .github/workflows/create-packages.yml:75]: Wow it got a lot simpler with your idea, thanks! :)

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: .github/workflows/draft-release.yml (hunk 1 above)
Comment (alltilla) [reply .github/workflows/draft-release.yml]: I had to remove the cancelled comment, as I cannot check the whole workflow's status, I would need to check each previous checks if they were cancelled manually, and write the necessary comment.

This comment just causes so much trouble....

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: .github/workflows/create-packages.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (gaborznagy) [reply .github/workflows/create-packages.yml:75]: Thanks for the change!

I would accept this version, but I would like to hear your opinion first. I've unresolved this thread for visibility reasons only.

I like the current version however, I think the additional `Prepare package for artifact` step could be skipped. 
I think it's unnecessary and the extra directory serves no meaning.

IMHO, We could simplify it:
```
      - name: Create package
        working-directory: syslog-ng
        run: |
          ./dbld/rules package-${{ matrix.distro }}

      - name: Store package as artifact
        uses: actions/upload-artifact@v3
        with:
          name: package-${{ matrix.distro }}
          path: syslog-ng/dbld/build/${{ matrix.distro }}
```

`upload-packages` workflow would work without changes, as it only cares about the name of the artifact name, and it doesn't care much about the internal structure of the artifact: it just downloads it to a given path, then uploads it to Azure.

File: .github/workflows/draft-release.yml (hunk 1 above)
Comment (gaborznagy) [reply .github/workflows/draft-release.yml]: I gave it another try and I think your latest patch is perfect for the job (or workflow... pun intended).

You are using the expression `if: (github.event.pull_request.merged && contains(github.event.pull_request.labels.*.name, 'version-bump'))` in the `comment-workflow-result` job, which is the same as used in initial job `create-release-tarball`.

Since the `draft-release` runs on every closed PR (either merged or not), we need some kind of filtering when to run the draft-release workflow. We had the very same filter before you refactored `draft-release`, and we need it for every job in the workflow.
I've checked github actions doc for an alternative, but haven't found a better way.

File: .github/workflows/create-packages.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (alltilla) [reply .github/workflows/create-packages.yml:75]: I think the problem with your solution is that when you explicitly give a directory path for `upload-artifact`, it will not upload the directory and its contents, but only its contents (or at least this is how I remember). I will check it again though, it would be nice to remove the preparation step.

[the remark this answers is not in the document]
Comment (gaborznagy) [reply .github/workflows/create-packages.yml:75]: Is it necessary to have a directory, though? The `upload-packages` step will download the artifact into a separate directory, why need another directory inside that?

[the remark this answers is not in the document]
Comment (alltilla) [reply .github/workflows/create-packages.yml:75]: When we upload to Azure, we need to put the deb files for example to `incoming/nightly/debian-buster/...deb`. So we need to have the `debian-buster` directory path somewhere.

The `azure-cli` step in `upload-packages` needs a `--source` (local) and a `--destination` (remote) parameter, and it will match the contents of the two. The way it works now is that the source is `package` (where the artifact is downloaded, currently has the `debian-buster` dir inside) and the destination is `/incoming/nightly`. I think it is clean this way, `upload-artifact` does not make tricks with the paths, it simply uploads the content of the artifact to `incoming/nightly/` for example. It would be possible to add `debian-buster` to the end of `--destination`, but then it would introduce some unnecesary logic to `upload-packages`, which might be better to have in `create-packages` instead.

I think having the necessary directory structure in the artifact itself, and not delegating this logic to `upload-packages` is cleaner in a high level perspective. Low level it introduces a "hacky" step, which is needed because of the way `upload-artifact` works. I would prefer this way.

[the remark this answers is not in the document]
Comment (alltilla) [reply .github/workflows/create-packages.yml:79]: Updated.

Thanks for the discussion :)

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Review by gaborznagy — APPROVED** [review]
Thanks for the changes!

**Comment by bazsi** [comment]
This looks AWESOME. :)

**Review by bazsi — APPROVED** [review]

**Review by MrAnno — APPROVED** [review]
